docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,338 @@
+This is similar to what Netlify does:
+https://docs.netlify.com/routing/redirects/redirect-options/#trailing-slash.
+
+Page and exact redirects without a wildcard at the end will be normalized,
+all other redirects need to be matched as is.
+
+This makes it impossible to match a path with a trailing slash.
+
+Explicit ``$rest`` placeholder
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Explicitly place the ``$rest`` placeholder in the target URL,
+instead of adding it automatically.
+
+Some times users want to redirect to a different path,
+we have been adding a query parameter in the target URL to
+prevent the old path from being added in the final path.
+For example ``/new/path/?_=``.
+
+Instead of adding the path automatically,
+users have to add the ``$rest`` placeholder in the target URL.
+For example:
+
+From:
+  ``/old/path/$rest``
+To:
+  ``/new/path/$rest``
+
+From:
+  ``/old/path/$rest``
+To:
+  ``/new/path/?page=$rest&foo=bar``
+
+Improving page redirects
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Allow to redirect to external domains.
+  This can be useful to apply a redirect of a well known path
+  in all versions to another domain.
+
+  For example, ``/security/`` to a their security policy page in another domain.
+
+  This new feature isn't strictly needed,
+  but it will be useful to simplify the explanation of the feature
+  (one less restriction to explain).
+
+- Allow a wildcard at the end of the from path.
+  This will allow users to migrate a whole directory to a new path
+  without having to create an exact redirect for each version.
+
+  Similar to exact redirects, users need to add the ``$rest`` placeholder explicitly.
+  This means that that page redirects are the same as exact redirects,
+  with the only difference that they apply to all versions.
+
+Improving Sphinx redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These redirects are useful, but we should rename them to something more general,
+since they apply to all types of projects, not just Sphinx projects.
+
+Proposed names:
+
+- HTML URL to clean URL redirect (``file.html`` to ``file/``)
+- Clean URL to HTML URL redirect (``file/`` to ``file.html``)
+
+Other ideas to improve redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Run forced redirects before built-in redirects.
+  We currently run built-in redirects before forced redirects,
+  this is a problem when moving a whole project to a new domain.
+  For example, a forced redirect like ``/$rest``,
+  won't work for the root URL of the project,
+  since ``/`` will first redirect to ``/en/latest/``.
+
+  But shouldn't be a real problem, since users will still need to
+  handle the ``/en/latest/file/`` paths.
+
+- Run redirects on the edge.
+  Cloudflare allow us to create redirects on the edge,
+  but they have some limitations around the number of
+  redirect rules that can be created.
+
+  And they will be useful for forced exact redirects only,
+  since we can't match a redirect based on the response of the origin server.
```

Comment:
I think this will open the door for a lot of complexity and duplicated code on Python and Javascript. I'd say that unless it's strictly required, I wouldn't follow this direction for now.

In my experience with CF workers, they are not super easy to test and debug and it's hard to do local testing. I'm sure this experience can be improved, but I'd stick with Python for now if possible.